### PR TITLE
Rollback handling implementation

### PR DESCRIPTION
**Pull request by CSubmerged**
Tidy version
This set implements rollback handling for the simple_dcp_client.py including persistence accompanied by log files.

### CODE REVIEW

File: lib/dcp_bin_client.py [hunk 1]
```diff
@@ -595,7 +596,7 @@ def formated_response(self, opcode, keylen, extlen, dtype, status, cas, body, op
                 by_seqno, rev_seqno, ext_meta_len = \
                     struct.unpack(">QQH", body[0:18])
 
-            print delete_time
+            # print delete_time
```
Comment (daverigby) [inline lib/dcp_bin_client.py]: Don't leave commented code committed; if you want to delete this just delete it :)

File: simple_dcp_client.py [hunk 2]
```diff
@@ -222,8 +221,23 @@ def parseArguments():
     return parser.parse_args()
 
 
+def argument_exceptions(args):
```
Comment (daverigby) [inline simple_dcp_client.py]: Style: it's not very clear to me from the name what this function does. While they are mostly about C++; the KV-Engine coding standards at https://github.com/couchbase/kv_engine/blob/master/docs/CodingStandards.rst#name-types-functions-variables-and-enumerators-properly give guidance on naming things. In this instance, consider a verb phrase for naming functions / methods.

File: lib/dcp_bin_client.py [hunk 3]
```diff
@@ -29,7 +29,7 @@ def __init__(self, host='127.0.0.1', port=11210, timeout=30, do_auth=True):
         self.collections = False
 
         # Option to print out opcodes received
-        self.__opcode_dump = False
+        self._opcode_dump = False
```
Comment (mcpaddy) [inline lib/dcp_bin_client.py]: Any reason why we went from superprivate to semiprivate?
I can't find example of using it outside of the class.

File: simple_dcp_client.py [hunk 4]
```diff
@@ -273,6 +293,8 @@ def parseArguments():
     parser.add_argument("--failover-log", help= "Option to input a custom failover log via a file path, in the form of \
     a list of tuples seperated by new lines: (uuid, seq.no) \n(uuid, seq.no)\n(uuid, seq.no)...",
                         required=False, type=str)
+    parser.add_argument("--log-reset", "-l", help="Disable initial reset of log files", required=False,
```
Comment (daverigby) [inline simple_dcp_client.py]: Consider your user interface here - I think having an option which disables deleting something breaks the principle of least surprise. 

It's also generally poor practice to silently create files (with a fixed name) in a code repository - and then have to add them to an ignore list.

I'd suggest flipping this logic - don't save failover logs by default; but add an option to allow the user to specify a path of where they should be loaded / saved from. That makes the usage much more explicit; and gives the user control of where files are written.

File: logs.py [hunk 5]
```diff
@@ -0,0 +1,94 @@
+# Functions for dealing with log updates, to persist uuids and seq_nos for each vbucket
```
Comment (daverigby) [inline logs.py]: Correct me if I'm wrong (my python fu is weak these days), but isn't the convention to use a block comment (""") for file-level comments?

File: logs.py [hunk 6]
```diff
@@ -0,0 +1,94 @@
...
+
+import json
+import os
+
+
+def get_path(vb):
```
Comment (daverigby) [inline logs.py]: Style: for non-trivial functions/methods consider even a brief docstring explaining what it does.

File: logs.py (hunk 5 above)
Comment (mcpaddy) [inline logs.py]: Is logs.py a good name? 
Should this be a part of dcp bin client?

File: logs.py [hunk 7]
```diff
@@ -0,0 +1,94 @@
...
+    dir_path = os.path.dirname(os.path.realpath(__file__))
+    path = str(dir_path) + '/logs/' + str(vb) + '.json'
```
Comment (daverigby) [inline logs.py]: Look at path.join() (I forget the exact name) as that works correctly on Win32.

File: logs.py [hunk 8]
```diff
@@ -0,0 +1,94 @@
...
+    if not os.path.exists(path):  # The file doesn't exist
```
Comment (daverigby) [inline logs.py]: As per my later comments about user interface; consider if it's a better experience to handle missing files as empty?

File: simple_dcp_client.py [hunk 9]
```diff
@@ -3,6 +3,8 @@
 import pprint
 import time
 import sys
+import logs
```
Comment (daverigby) [inline simple_dcp_client.py]: `logs` is somewhat of an ambiguous module name (IMHO); maybe make it more explicit to people don't perhaps confuse this with a system package?

File: logs.py [hunk 10]
```diff
@@ -0,0 +1,94 @@
...
+    for vb in vb_list:
+        path_string = get_path(vb)
+        with open(path_string, 'w') as f:
+            json.dump({}, f)
+
+
+def upsert_failover(vb, failover_log):
+    path_string = get_path(vb)
+
+    with open(path_string, 'r') as vb_log:
```
Comment (mcpaddy) [inline logs.py]: Is there a good reason to do two opens on the same file?

File: logs/foo.txt [hunk 11]
```diff
@@ -0,0 +1 @@
+Placeholder file to ensure that folder exists within git.
```
Comment (mcpaddy) [inline logs/foo.txt]: Surely that is the application tasks and not the source code repos job?

File: lib/dcp_bin_client.py (hunk 3 above)
Comment (CSubmerged) [reply lib/dcp_bin_client.py]: Was attempting to follow the convention used within the bin_client files. Seeming as the opcode dump variable is controlled via simple_dcp_client I assumed having it not superprivate might allude to that, but otherwise no reason.

File: simple_dcp_client.py [hunk 12]
```diff
@@ -127,16 +140,15 @@ def initiate_connection(args):
...
+
     global dcp_client
     dcp_client = DcpClient(host, int(port), timeout=timeout, do_auth=False)
     print 'Connected to:', node
 
     try:
         response = dcp_client.sasl_auth_plain(args.user, args.password)
     except MemcachedError as err:
-        print err
+        print err, 'error'
```
Comment (mcpaddy) [inline simple_dcp_client.py]: What's the point on postfixing  'error'?

**Comment by CSubmerged** [comment]
Looking back I agree @daverigby with the comments on the commit summaries, I'm putting too much of the needed info in the description which leaves the main summary vague. It's a little difficult to go back and edit the commit names, though I would suggest to change it to 'Send bug fix and console clarity updates' and 'Server req info and UUID argument exceptions' respectively. In terms of this request as a whole, would it be suitable to squash the commits?

File: simple_dcp_client.py [hunk 13]
```diff
@@ -203,6 +208,45 @@ def add_streams(args):
...
+    # If argument to use manual log
+    if args.failover_log:
+        f_log = open(args.failover_log, 'r')
+        failover_values = []
+        for line in f_log:
+            line = line.replace('(', '')
+            line = line.replace(')', '')
+            split = line.split(',')
+            value = tuple([int(split[0]), int(split[1])])
+            failover_values.append(value)
```
[the remark this answers is not in the document]
Comment (CSubmerged) [reply simple_dcp_client.py]: There's a strong case for saying that option is made completely obsolete with the introduction of editable logs that can handle multiple vbuckets in JSON format. For this reason, I'm planning on removing that argument option entirely in favour of using the log format.

File: logs.py (hunk 5 above)
Comment (CSubmerged) [reply logs.py]: logs.py is a pretty bad name, though there is an argument for having it separate from inside dcp_bin_client due to the scope of files it deals with. dcp_bin_client is also imported by a number of other python scripts within the repo, and I have no guarantee that the functions in logs.py will work / be useful for those.

File: simple_dcp_client.py (hunk 9 above)
Comment (CSubmerged) [reply simple_dcp_client.py]: Good point, also brought up by @mcpaddy, has been changed to dcp_data_persist for clarity and also potential future extension for other vbucket data.

**Comment by daverigby** [comment]
> Looking back I agree @daverigby with the comments on the commit summaries, I'm putting too much of the needed info in the description which leaves the main summary vague. It's a little difficult to go back and edit the commit names, though I would suggest to change it to 'Send bug fix and console clarity updates' and 'Server req info and UUID argument exceptions' respectively. In terms of this request as a whole, would it be suitable to squash the commits?

Feel free to squash / re-base changes - certainly for KV-Engine core code we don't use a GitHub-style branch merge model (I find it pretty hard to follow what I actually need to still review here / what's changed by just stacking more "fixup" commits!)

**Comment by CSubmerged** [comment]
@daverigby The commits have now been rebased, cleaned and should make a lot more sense in general (hopefully!), so please take a look if you have the chance

File: dcp_data_persist.py [hunk 14]
```diff
@@ -0,0 +1,159 @@
...
+        if self.internal:
+            if str(vb) in self.dictstore.keys():
+                self.dictstore[str(vb)]['failover_log'] = failover_log
+            else:
+                self.dictstore[str(vb)] = {'failover_log': failover_log}
+
+        else:
+            path_string = self.get_path(vb)
+
+            with open(path_string, 'r') as vb_log:
```
Comment (jameseh96) [inline dcp_data_persist.py]: This may be fine as is, but it might be worth being aware that as the log grows this will lead to a lot of write amplification - I've not absorbed the wider context yet, but if this is expected to be called regularly it might be worth caching the changes and flushing many together? I don't know if that would be practical here though :)

File: dcp_data_persist.py [hunk 15]
```diff
@@ -0,0 +1,159 @@
...
+                data = json.load(vb_log)
+
+            data['failover_log'] = failover_log
+
+            with open(path_string, 'w') as vb_log:
+                json.dump(data, vb_log)
+
+    def upsert_sequence_no(self, vb, seq_no):
+        """ Insert / update sequence number, and move old sequence number to appropriate list """
+        if self.internal:
```
Comment (jameseh96) [inline dcp_data_persist.py]: This looks to be becoming a pattern - might it be worth refactoring the actual fetching and storing out to a couple of methods that check `if self.internal`? then each method that wants to log could (hypothetically) just do

```
mydata = load(blah)
mydata["thing"].append("blah")
store(mydata)
```

(just a general idea, not specific code suggestion :P )

without having to care where or how it is stored? Again, you'll know better how workable that would be.

File: dcp_data_persist.py [hunk 16]
```diff
@@ -0,0 +1,159 @@
...
+            for vb in vb_list:
+                path_string = self.get_path(vb)
+                print path_string
+                with open(path_string, 'w') as f:
+                    json.dump({}, f)
+
+    def upsert_failover(self, vb, failover_log):
+        """ Insert / update failover log """
+        if self.internal:
+            if str(vb) in self.dictstore.keys():
```
Comment (jameseh96) [inline dcp_data_persist.py]: You might want to take a look at `collections.defaultdict`, saves doing this kind of check

File: dcp_data_persist.py [hunk 17]
```diff
@@ -0,0 +1,159 @@
...
+
+        return read_dict
+
+    def get_seq_nos(self, vb_list):
+        """ Return a dictionary where keys are vbuckets and the data is the sequence number """
+        read_dict = {}
+
+        if self.internal:
+            for vb in vb_list:
+                if str(vb) in self.dictstore.keys():
```
Comment (jameseh96) [inline dcp_data_persist.py]: IIRC, you can do

`if "some_key" in some_dict:`

without .keys()

Might want to see if there is a PEP 8 style guide thing on what is preferred?

File: dcp_data_persist.py [hunk 18]
```diff
@@ -0,0 +1,159 @@
...
+                    read_dict[str(vb)] = self.dictstore[str(vb)].get('seq_no')
```
Comment (jameseh96) [inline dcp_data_persist.py]: OOI, why are vbs used as strings so much, but (presumably) not stored as them? would it be easier to uniformly keep them as one type?

File: dcp_data_persist.py [hunk 19]
```diff
@@ -0,0 +1,159 @@
...
+    to persist uuids and seq_nos for each vbucket.
+    Currently stored within folder called 'logs'
+    Potential scope to save more data about vbucket if needed
+"""
+
+import json
+import os
+
+
+class LogData:
```
[the remark this answers is not in the document]
Comment (CSubmerged) [reply dcp_data_persist.py]: https://www.python.org/doc/newstyle/ Yep seems so

File: dcp_data_persist.py [hunk 20]
```diff
@@ -0,0 +1,159 @@
...
+        else:
+            """ Create with internal file logging """
+            self.internal = True
+            self.dictstore = {}
+
+    def get_path(self, vb):
+        """ Retrieves path to log file for inputted virtual bucket number """
+        # Make directory if it doesn't exist
+        if not self.internal:
+            fullpath = os.path.join(self.path, os.path.normpath(str(vb) + '.json'))
```
[the remark this answers is not in the document]
Comment (CSubmerged) [reply dcp_data_persist.py]: I'm not sure it entirely matters in Python https://stackoverflow.com/questions/38722105/format-strings-vs-concatenation/43213810

File: dcp_data_persist.py (hunk 15 above)
Comment (jameseh96) [reply dcp_data_persist.py]: Related to the caching suggestion - if caching were to be implemented this would likely be made neater in the process - one could imagine always storing it as if `self.internal == True` (avoiding having this branch in every method) and then just having a `flush()` method or something that actually writes it to disk if a path was given

File: dcp_data_persist.py (hunk 20 above)
[the remark this answers is not in the document]
Comment (jameseh96) [reply dcp_data_persist.py]: From that link
    It's just for the looks

Good looking Python is the best Python ;)
it also avoids `str(vb)`, format will take care of that.

**Comment by jameseh96** [comment]
Also, obligatory commit message advice link: https://chris.beams.io/posts/git-commit/

File: dcp_data_persist.py (hunk 20 above)
[the remark this answers is not in the document]
Comment (jameseh96) [reply dcp_data_persist.py]: Also OOI, why are you `normpath`ing? can `vb` contain slashes?

File: dcp_data_persist.py [hunk 21]
```diff
@@ -0,0 +1,159 @@
...
+
+    def reset(self, vb_list):
+        """ Clears/makes files for list of virtual bucket numbers"""
+        if self.internal:
+            self.dictstore = {}
+        else:
+            for vb in vb_list:
+                path_string = self.get_path(vb)
+                print path_string
+                with open(path_string, 'w') as f:
```
[the remark this answers is not in the document]
Comment (CSubmerged) [reply dcp_data_persist.py]: Not sure how that would be neater. The reset function is there to prepare the log files for new entry, not remove them. By opening them as 'w' it clears whatever was there if there was something. Although I can't think of a use for it in simple_dcp_client at the mo, it may be worth implementing a delete function just to clarify it

File: dcp_data_persist.py (hunk 15 above)
Comment (CSubmerged) [reply dcp_data_persist.py]: I guess the idea of having internal everywhere in this file is to accomplish exactly that for the simple_dcp_client file instead. Will have a look into it though

File: dcp_data_persist.py (hunk 18 above)
Comment (CSubmerged) [reply dcp_data_persist.py]: From the original document arguments, vbuckets are defined as int's inside a list, so I've mainly been working off that and assumed that ensuring it's a string is a little bit safer.

File: dcp_data_persist.py (hunk 14 above)
Comment (jameseh96) [reply dcp_data_persist.py]: Troll suggestion: copy Couchstore and have an append only file of JSON objects storing only the changes and then run compaction every so often ;)

File: dcp_data_persist.py (hunk 20 above)
[the remark this answers is not in the document]
Comment (CSubmerged) [reply dcp_data_persist.py]: Was using normpathing to originally ensure that the extension of logs/{}.json ended up as a valid path to join. When the logs/ part got moved to the init of the class though, I hadn't considered the need to change it. vb cannot contain slashes as it is fixed to a number (0-1023), and if the user inputs something else it'll break somewhere else in the program first!

File: simple_dcp_client.py [hunk 22]
```diff
@@ -3,6 +3,8 @@
 import pprint
```
Comment (jameseh96) [inline simple_dcp_client.py]: I know these imports precede your changes, but would you be able to run "optimise imports" (if you're in JetBrains land) - should order them PEP 8 ishly 👍

File: simple_dcp_client.py [hunk 23]
```diff
@@ -217,13 +291,41 @@ def parseArguments():
     parser.add_argument("--noop-interval", help="Set time in seconds between NOOP requests", required=False)
     parser.add_argument("--opcode-dump", help="Dump all the received opcodes via print", required=False,
                         action="store_true")
+    parser.add_argument("--stream-req-info", help="Display vbuckets, seq no's and uuid with every stream request",
+                        required=False, action="store_true")
+    parser.add_argument("--uuid", help="Set the vbucket UUID", type=int, default=0, required=False)
+    parser.add_argument("--failover-logging", help="Enables use of persisted log JSON files for each vbucket, which \
+    contain the failover log and sequence number", required=False, action='store_true')
+    parser.add_argument("--log-path", help="Set the file path to use for the log files", default=None, required=False)
+    parser.add_argument("--log-preset", "-l", help="Enable initial preset of log files", required=False,
```
Comment (jameseh96) [inline simple_dcp_client.py]: To confirm I've understood, --log-preset is essentially "no clobber"?

File: simple_dcp_client.py [hunk 24]
```diff
@@ -217,13 +291,41 @@ def parseArguments():
...
     parser.add_argument("-p", "--password", help="Password", required=True)
-    return parser.parse_args()
+    parsed_args = parser.parse_args()
+    if (parsed_args.log_path or parsed_args.log_preset) and not parsed_args.failover_logging:
+        parser.error("Both --log-path and --log-preset require --failover-logging to function.")
+    return parsed_args
+
+
+def convert_special_argument_parameters(args):
+    if args.vbuckets == ['-1']:
```
Comment (jameseh96) [inline simple_dcp_client.py]: I'm being super blind/under coffeed - could you point me to where args.vbuckets comes from? I can't see a parser argument that sets vbuckets as its destination.

File: simple_dcp_client.py (hunk 23 above)
Comment (CSubmerged) [reply simple_dcp_client.py]: Pretty much yes. During the normal file setup part it will check if that log file exists, and this option stops it from overwriting that. However, if it doesn't exist it will still create a blank file for it to use. (ngl hadn't heard of clobber before, hopefully https://en.wikipedia.org/wiki/Clobbering is mostly accurate)

File: simple_dcp_client.py (hunk 13 above)
[the remark this answers is not in the document]
Comment (dhaikney) [reply simple_dcp_client.py]: Yeah, ^ that's fairly horrific :) use JSON.

[the remark this answers is not in the document]
Comment (CSubmerged) [reply simple_dcp_client.py]: True. It get's replaced by JSON in the following commit, though I believe it remained in this commit to mark out the functionality.
